ROADMAP.md

```diff
@@ -0,0 +1,595 @@
+
+**Why?** It's a crucial requirement in our visionary goal of
+decentralising networks. WIP.
+
+**Links:**
+
+-   [On \"DialMe\", NAT traversal, Relay, line switching, packet switching
+    (libp2p protocols)](https://github.com/ipfs/notes/issues/110)
+
+## Evolve
+
+**Our short-term roadmap**.
+
+This is the stuff pushing the existing libp2p stack forward.
+
+### 🤝 Low latency, efficient connection handshake
+
+**High priority for: IPFS**
+
+**What?** Establishing a connection and performing the initial handshake
+should be as cheap and fast as possible. Supporting things like
+*selective* stream opening, *preemption*, *speculative* negotiation,
+*upfront* negotiation, protocol table *pinning*, etc. may enable us to
+achieve lower latencies when establishing connections, and even the
+0-RTT holy grail in some cases. These features are being discussed in
+the *Protocol Select* protocol design.
+
+**Why?** Multistream 1.0 is chatty and naïve. Streams are essential to
+libp2p, and negotiating them is currently inefficient in a number of
+scenarios. Also, bootstrapping a multiplexed connection is currently
+guesswork (we test protocols one by one, incurring in significant
+ping-pong).
+
+**Links:**
+
+- [Multiselect 2.0 design doc](https://github.com/libp2p/specs/pull/205).
+
+- [Multiselect 2.0 specification](https://github.com/libp2p/specs/pull/227)
+
+- [RFC Multistream 2.0](https://github.com/libp2p/specs/pull/95).
+
+### 🕸 Unprecedented global connectivity
+
+**What?** A DHT crawl measurements (Nov 22nd 2019) showed that out
+of 4344 peers, 2754 were undialable (\~63%). This evidence correlates
+with feedback from the IPFS and Filecoin teams.
+
+We need to implement additional mechanisms for Firewall and NAT traversal to
+have the highest probability of being able to establish a direct connections.
+Mechanisms we wish to add include:
+
+- Project Flare stack (via *Circuit Relay v2*, *Direct Connection Upgrade
+  through Relay*, *AutoNAT*, *Stream Migration*, ...)
+
+- WebRTC
+
+**Why?** Good connectivity is the bread-and-butter of libp2p. Focusing
+on solving these issues will bring more stability and robustness to the
+rest of the system.
+
+**Links:**
+
+- [Hole punching long-term
+  vision](https://github.com/mxinden/specs/blob/hole-punching/connections/hole-punching.md).
+
+- [NAT traversal tracking issue](https://github.com/libp2p/specs/issues/312).
+
+### Standardized interaction of different protocols via Smart Records
+
+TODO: Extend
+
+Copied and adapted from: https://github.com/libp2p/go-smart-record/blob/4a1b33c6d7cb2fb38d166abc1023fbbbab33b60f/README.md
+
+**What?** With Smart Records (SRs) we generalize DHT's key/value put/get as a
+separate protocol that can be leveraged by any other protocol (including DHT
+protocols) to store arbitrary data.
+
+**Why?** We currently don't have a standardized, shared, public medium decoupled
+from the transport protocol for the interaction of different protocols.
+
```

Comment:
```suggestion
### Composable routing

Composable Routing is a framework that was designed to enable the
interoperability of protocols and nodes with different versions and capabilities,
developed by independent teams or organizations. As a consequence, it also
enables middleware components (like caching, batching, predictive fetching, and so on)
and the co-existence of different content finding protocols and subnetworks.

The framework has four "pillars":
* Routing syntax and language
* Routing interface
* Smart Records
* Subnets

#### Routing syntax and language

The routing syntax is a common data model (with a user-facing syntactic representation)
for communicating routing-related information (requests and responses).
The [implementation of the routing syntax](https://github.com/libp2p/go-routing-language/tree/master/syntax)
provides serialization and pretty printing of routing expressions.

The routing language is a semantic that specifies how to interpret
routing expressions, and specifically how to compose
complex routing statements out of simpler ones.

Finally, the routing grammar is a set of rules that define the "vocabulary"
of the language. In other words, the grammar defines the set of 
valid and recognizable routing statements. The grammar is, by design,
a continuously evolving specification of the recognizable language vocabulary.

The grammar is reflected in [code](https://github.com/libp2p/go-routing-language/tree/master/patterns),
as a collection of parsing functions for the valid statements in the language. When new rules
are coined (and added to the grammar specification), their respective parsers are codified manually. Going forward,
the project plans to have a grammar compiler, which generates parsers
and checks for ambiguities.

For intuition, expressions in the routing language can represent statements, like:
* I am providing CID via bitswap.
* I am interested in having the content of CID.
* Fetch the content corresponding to IPFS_PATH.

They can also express compositions of statements, like:
* Find providers for CID, then fetch its content,
then extract the field "foo" and view it as a CID,
then fetch its content.

#### Routing interface

A routing system processes routing requests and returns routing responses.
Routing requests and responses have semantics, which are unlike those of a
traditional request/response protocols, that implement function call arguments
and return values.

A routing request is a routing expression which represents a program to be evaluated
by the routing system. In the general case, a routing system may not be able to
evaluate all steps in a program, since it may have limited capabilities or a legacy
version which does not understand newer routing vocabulary. For this reason,
routing responses have to follow a format that allows the routing system to
communicate back the results of partially evaluating the routing request.

Specifically, a routing response is definitionally a copy of the routing request
with the evaluated sub-expressions substituted by the result of their evaluation
or by equivalent expressions (expressions that would evaluate to the same result).

In other words, a routing request and a response are both (usually different)
programs which, if evaluated fully, would produce the same result. In the 
realm of lambda calculus and formal verification, this is known as beta-equivalence.

#### Smart records

Smart Records (SR) is a technology that enables multiple decentralized parties and protocols
to randevouz and share information on a topic. Specifically, SR is a key-value store
service, which facilitates conflict-free updates to values by multiple writers.

SR keys are opaque strings, while SR values are hierarchical documents in the
data model of the routing syntax. An SR service provides a value update operation,
which enables a writer to update a subexpression of a value without interfering with
the rest of the value's contents. Update operations are commutative by design,
ensuring that (i) operation re-ordering due to asynchronous phenomena has no
effect on semantics, and (ii) multiple physical replicas of a value can be merged
unambiguously. In other words, SR values are CRDTs.

The quintessential application of SR is to use it as the store for DHT put/get operations.
This enables any third-party protocol to use the DHT for its own purposes, respectively
unlocking a vast diversity of applications that can be developed by the community.
Some simple examples are email-over-DHT and chat-over-DHT.

More generally, a DHT with flexible value semantics is effectively a giant 
general-purpose, random-access storage layer. And, therefore, highly complex
applications, like a crypto exchange-over-DHT, could be built on this substrate.

The current [implementation](https://github.com/libp2p/go-smart-record) of the SR
virtual machine supports in-memory storage only. We hope to add support for persistence.
The SR virtual machine can also be used programmatically as a Go library that implements a
key-value store for routing expressions. This use comes up, for instance, when implementing
a filecoin content indexer, which must store multiple different provide records for every cid.

#### Subnets

At present, there is a single DHT network which spans across all IPFS nodes.
We would like to introduce a general standard for the co-existence of
multiple network instances with individual discovery, membership and content routing semantics.

This would enable communities of users to configure custom networks
that meet their physical, business, logistics, privacy, security and other
application needs.

Subnets address this goal. A subnet is effectively a set of peers that run
an instance of a routing protocol (e.g. DHT or pubsub) amongst themselves.
Subnets can have custom membership and authentication semantics.
Optionally, a subnet can have an externally-facing routing interface, allowing
others to use it as a routing service.

```

Reply:
Thanks @petar for the detailed suggestion. Given that each section is only giving a short overview, I added a reduced version in https://github.com/libp2p/specs/pull/339/commits/64a7be3fe1cd9de39fe38845f031596b24424959 linking to the full description in https://github.com/libp2p/specs/issues/343. 

Let me know what you think.

Reply:
Resolving here as "Composable Routing" section has been included.